File: packages/types/src/recurring.ts
import { z } from 'zod';

import { createExpenseSchema } from './expenses';

/** Mirrors the Prisma RecurringFrequency enum — kept in sync manually, same as SplitType/GroupType. */
export const recurringFrequencySchema = z.enum(['DAILY', 'WEEKLY', 'MONTHLY', 'YEARLY']);
export type RecurringFrequency = z.infer<typeof recurringFrequencySchema>;

/**
 * ABRO_PRD.md §35: "Each recurring expense contains: template, frequency,
 * next execution, enabled." The template IS a real Expense (created via the
 * normal expense-create path, per docs/BACKEND_PLAN.md item 4 -- "reuse,
 * don't duplicate split computation"), so creating a recurring expense takes
 * every CreateExpenseInput field plus `frequency`. `expenseDate` doubles as
 * the template's date and the anchor `nextRunAt` is computed from.
 */
export const createRecurringExpenseSchema = z.intersection(
  createExpenseSchema,
  z.object({ frequency: recurringFrequencySchema }),
);
export type CreateRecurringExpenseInput = z.infer<typeof createRecurringExpenseSchema>;

export const setRecurringEnabledSchema = z.object({ enabled: z.boolean() });
export type SetRecurringEnabledInput = z.infer<typeof setRecurringEnabledSchema>;

File: apps/api-go/internal/apitypes/recurring.go
package apitypes

import (
	"time"

	"github.com/Substance-k3n/abro/apps/api/internal/httpx"
)

var validFrequencies = map[string]bool{"DAILY": true, "WEEKLY": true, "MONTHLY": true, "YEARLY": true}

// CreateRecurringExpenseInput mirrors packages/types' createRecurringExpenseSchema
// -- every CreateExpenseInput field plus frequency (ABRO_PRD.md §35: the
// template IS a real Expense).
type CreateRecurringExpenseInput struct {
	CreateExpenseInput
	Frequency string `json:"frequency"`
}

func (in *CreateRecurringExpenseInput) Validate() error {
	if err := in.CreateExpenseInput.Validate(); err != nil {
		return err
	}
	if !validFrequencies[in.Frequency] {
		return httpx.BadRequest("VALIDATION_ERROR", "frequency must be one of DAILY, WEEKLY, MONTHLY, YEARLY")
	}
	return nil
}

type SetRecurringEnabledInput struct {
	Enabled bool `json:"enabled"`
}

type RecurringExpense struct {
	ID              string      `json:"id"`
	Frequency       string      `json:"frequency"`
	NextRunAt       time.Time   `json:"nextRunAt"`
	Enabled         bool        `json:"enabled"`
	CreatedAt       time.Time   `json:"createdAt"`
	UpdatedAt       time.Time   `json:"updatedAt"`
	TemplateExpense AuthExpense `json:"templateExpense"`
}

type GenerateDueResult struct {
	Generated int `json:"generated"`
}
